Tests optimisations, expose task properties and cleaning up temporary files

piotrminkina opened the pull request:
First of all, I have significantly improved test performance. Executing the tests performs from 6 to as long as 13 minutes (it probably depended on the load on the Gradle server, etc.). The first performance optimization was to separate a common directory where Gradle distributions were downloaded. This change alone reduced test execution time for me in the range from 2.5 to 3.5 minutes.

The next optimization was to disable deleting the directory with downloaded Gradle distributions, but in order not to clutter the developer's disk, I changed the directory location to a temporary one for the "Test" task. Executing `./gradlew clean` in the Project will also clear these files. This change has resulted in each subsequent test execution now taking only about 45 seconds. Much better and faster feedback than original!

I noticed that both the plug-in and the tests leave a lot of junk in the `/tmp` directory, which multicollectively led to disk space exhaustion in my case. I addressed this problem by setting aside a shared temporary directory, which is created and deleted by Jupiter.

In addition to the above, I have made changes that address issues #7 and #8. I myself badly need these changes to minimize configuration effort and to get easy access to generated custom Gradle distributions along with the ability to add libraries and plugins (including my own) as early as the build stage. I also added the missing `.gitignore` file with a very minimal set of exclusions.

I will be very grateful for the smooth integration of my changes into your project.

Commit:
Add missing .gitignore file for project contributors

Commit:
Exposed BuildCustomGradleDistributionTask inputs and outputs for more flexibility and better dependencies resolution (closes #8)

Commit:
Set project version as convention for customDistributionVersion (closes #7)

Commit:
Implemented configurable Gradle download directory and improve tests performance

Commit:
Cleaning temporary files that the plugin and tests leaves after finish

Commit:
Implement sharing Gradle downloads for much faster tests result feedback

piotrminkina committed:
Use Gradle conventions as defaults instead of own calculations

denis-zhdanov commented:
hi @piotrminkina, thank you for your interest and contribution!

Could you clarify how exactly did this change improve tests performance - `The first performance optimization was to separate a common directory where Gradle distributions were downloaded`? The test harness was already creating an empty dummy zip file, the plugin was not expected to download anything during the test.

I re-checked it once again and it seems that there is a place for improvement - currently we use `bin` base distribution type by default and prepare only one such dummy zip during the tests. However, there are some tests which target `all` distribution - the whole distribution is indeed downloaded during that. It's worth to improve that, but a proper way to fix that is just to generate the both dummy `all` and `bin` distributions in tests. That way we don't introduce unnecessary changes into production code and `build.gradle.kts` for tests

denis-zhdanov commented:
```
The next optimization was to disable deleting the directory with downloaded Gradle distributions, but in order not to clutter the developer's disk, I changed the directory location to a temporary one for the "Test" task. Executing ./gradlew clean in the Project will also clear these files. This change has resulted in each subsequent test execution now taking only about 45 seconds. Much better and faster feedback than original!
```
Looks like it is relevant to the previous comment - creating dummy `-all.zip` base distribution file by test harness achieves the same result

denis-zhdanov commented:
```
The next optimization was to disable deleting the directory with downloaded Gradle distributions, but in order not to clutter the developer's disk, I changed the directory location to a temporary one for the "Test" task. Executing ./gradlew clean in the Project will also clear these files. This change has resulted in each subsequent test execution now taking only about 45 seconds. Much better and faster feedback than original!
```
Looks like it is relevant to the previous comment - creating dummy `-all.zip` base distribution file by test harness achieves the same result

denis-zhdanov commented:
```
I noticed that both the plug-in and the tests leave a lot of junk in the /tmp directory, which multicollectively led to disk space exhaustion in my case. I addressed this problem by setting aside a shared temporary directory, which is created and deleted by Jupiter.
```

The intention was for `deleteOnExit()` to take care of that. Checked that in my local environment it's also not always automatically cleaned up. Agree that it's better to use standard junit mechanism for that

denis-zhdanov commented:
```
In addition to the above, I have made changes that address issues https://github.com/harmonysoft-tech/custom-gradle-dist-gradle-plugin/issues/7 and https://github.com/harmonysoft-tech/custom-gradle-dist-gradle-plugin/issues/8. I myself badly need these changes to minimize configuration effort and to get easy access to generated custom Gradle distributions along with the ability to add libraries and plugins (including my own) as early as the build stage. I also added the missing .gitignore file with a very minimal set of exclusions.
```
Looks good at the first glance, need to read about gradle task conventions though (they didn't exist when the plugin was created :) )

denis-zhdanov commented on src/main/kotlin/tech/harmonysoft/oss/gradle/dist/BuildCustomGradleDistributionTask.kt:
```diff
@@ -73,7 +104,7 @@ abstract class BuildCustomGradleDistributionTask : DefaultTask() {
         val gradleBaseName = "gradle-${extension.gradleVersion.get()}"
         val gradleZip = "$gradleBaseName-${extension.gradleDistributionType.get()}.zip"
-        val baseGradleArchive = project.layout.buildDirectory.file("download/$gradleZip").get().asFile
+        val baseGradleArchive = downloadRootDir.map { it.dir(gradleZip) }.get().asFile
```
why do we use `dir()` here? `file()` is more natural

piotrminkina replied:
You're right, apparently I was too hasty. I fixed it.

denis-zhdanov commented on src/main/kotlin/tech/harmonysoft/oss/gradle/dist/BuildCustomGradleDistributionTask.kt:
```diff
@@ -126,7 +153,8 @@ abstract class BuildCustomGradleDistributionTask : DefaultTask() {
 
     private fun getDistributions(): Collection<String> {
-        val childDirectories = extensionsRootDir.listFiles(FileFilter { it.isDirectory })
+        val extensionsRootDir = this.extensionsRootDir.orNull?.asFile
```
it's worth to fail with corresponding error message if this directory doesn't exist - the plugin is pointless otherwise

piotrminkina replied:
I forced `extensionsRootDir` to be required and the indicated directory to exist (commit df5b453). Gradle would return the correct error message to the user if the directory did not exist. For all test cases to succeed, I had to add two init.d directories with the `.gitkeep` file.

```quote
* What went wrong:
A problem was found with the configuration of task ':buildGradleDist' (type 'BuildCustomGradleDistributionTask').
  - In plugin 'tech.harmonysoft.oss.custom-gradle-dist-plugin' type 'tech.harmonysoft.oss.gradle.dist.BuildCustomGradleDistributionTask' property 'extensionsRootDir' specifies directory '/tmp/junit8917516433528672602/input7686248079884748448/src/main/resources/init.d' which doesn't exist.
    
    Reason: An input file was expected to be present but it doesn't exist.
    
    Possible solutions:
      1. Make sure the directory exists before the task is called.
      2. Make sure that the task which produces the directory is declared as an input.
    
    For more information, please refer to https://docs.gradle.org/8.4/userguide/validation_problems.html#input_file_does_not_exist in the Gradle documentation.
```

denis-zhdanov commented on src/main/kotlin/tech/harmonysoft/oss/gradle/dist/BuildCustomGradleDistributionTask.kt:
```diff
@@ -126,7 +153,8 @@ abstract class BuildCustomGradleDistributionTask : DefaultTask() {
+        val extensionsRootDir = this.extensionsRootDir.orNull?.asFile
+        val childDirectories = extensionsRootDir?.listFiles(FileFilter { it.isDirectory })
         return if (childDirectories == null || childDirectories.size < 2) {
             project.logger.lifecycle("using a single custom gradle distribution")
```
it's worth to verify that there is at least one file to be included into `init.d` in the resulting custom distribution. Otherwise the whole plugin is pointless

piotrminkina replied:
Is a change that forces the existence of the directory pointed to by `extensionsRootDir` sufficient?

denis-zhdanov replied:
sorry, what do you mean? My suggestion is to fail with a descriptive error if the directory doesn't exist

piotrminkina replied:
I mean this solution: https://github.com/harmonysoft-tech/custom-gradle-dist-gradle-plugin/pull/15#discussion_r1573696585. Gradle will fail when dir `src/main/resources/init.d` doesn't exist. Se commit df5b453.

denis-zhdanov replied:
sorry, don't see any check that if `init.d` directory is empty, we fail with a descriptive message. Did I miss something?

denis-zhdanov replied:
anyway, you already did a great contribution here, I will merge the PR now and add the change myself.
Thank you a lot!

denis-zhdanov commented:
finished review iteration

piotrminkina committed:
Remove unnecessary configuring Gradle download dir via project properties and optimize preparing fake downloads during tests

piotrminkina commented:
> it seems that there is a place for improvement - currently we use `bin` base distribution type by default and prepare only one such dummy zip during the tests.

OK, I improved the test implementation and now it is so that in the test case I specify which downloaded Gradle distribution to prepare. Now the tests take only 10 seconds to execute. Yay!

piotrminkina commented:
> The intention was for `deleteOnExit()` to take care of that. Checked that in my local environment it's also not always automatically cleaned up. Agree that it's better to use standard junit mechanism for that

This probably has to do with the fact that Gradle is running in daemon mode and therefore this function is not called. On the other hand, however, I was executing the `./gradlew --stop` command and this garbage still remained, so I don't quite know what was the reason for not calling this function.

piotrminkina committed:
Fixed call to generate path to Gradle distribution file

denis-zhdanov commented on src/test/kotlin/tech/harmonysoft/oss/gradle/dist/CustomGradleDistributionPluginTest.kt:
```diff
@@ -521,17 +570,24 @@ class CustomGradleDistributionPluginTest {
+        verify(testFiles.expectedRootDir, File(testFiles.inputRootDir, buildDistsDir))
+    }
+
+    private fun withDistributionType(type: String): CustomGradleDistributionPluginTest {
```
we can just create the both dummy `-bin.zip` and `-all.zip` in tests and not bother with explicitly specifying the target type. Less code - easier to maintain

piotrminkina replied:
At your request, I just simplified it :)

denis-zhdanov commented on src/main/kotlin/tech/harmonysoft/oss/gradle/dist/BuildCustomGradleDistributionTask.kt:
```diff
@@ -13,27 +13,55 @@ import java.nio.file.Files
-    private val includeRootDir: File
-        get() = project.file("src/main/resources/include")
+    @get:Internal
+    abstract val downloadRootDir: DirectoryProperty
```
do we still need this property? It's used only in tests

piotrminkina replied:
Yes, I use this in my multiproject, where the project structure is as below, and I have this property configured so that gradle distributions are downloaded to the `build/download` directory in the root project. This makes rebuilding individual custom distributions much faster.

```
build/
dists/
    8.2.1-bin/
        build/
        src/
    8.2.1-all/
        build/
        src/
```

denis-zhdanov commented:
finished review iteration

piotrminkina committed:
Made extensionsRootDir required

piotrminkina committed:
Simplified test implementation for creating fake downloaded Gradle distributions

piotrminkina committed:
Refactored configuring interface and annotate all public properties with task annotations to instrument Gradle when should rebuild custom distributions

piotrminkina committed:
Update samples to follow current plugin implementation

piotrminkina commented:
@denis-zhdanov With commit 5ee339c7, I refactored the configuration interface for the extension and task of your plugin. I think the properties names used now are more intuitive. Additionally, I annotated all extension properties with the appropriate annotations for the task. Now Gradle will know when to rebuild own distributions and when not to. I have updated the documentation with the current configuration capabilities of the plugin.

With commit 0e07eee6, I updated the samples to work with the current implementation of the plugin. Even without my changes, these examples did not work with the latest released version of the plugin.

piotrminkina commented:
Rebased with master

denis-zhdanov commented:
@piotrminkina please note that I was talking about this change - https://github.com/harmonysoft-tech/custom-gradle-dist-gradle-plugin/blob/master/src/main/kotlin/tech/harmonysoft/oss/gradle/dist/BuildCustomGradleDistributionTask.kt#L381

piotrminkina commented:
> @piotrminkina please note that I was talking about this change - https://github.com/harmonysoft-tech/custom-gradle-dist-gradle-plugin/blob/master/src/main/kotlin/tech/harmonysoft/oss/gradle/dist/BuildCustomGradleDistributionTask.kt#L381

While I understand what you meant, at the time I was asking whether the mere verification that the `init.d` directory exists, which Gradle does, would be sufficient. Well, but you implemented what you wanted and it is very cool.
